metrics-core/src/main/java/com/yammer/metrics/core/MetricName.java

```diff
@@ -205,29 +212,30 @@ public boolean equals(Object o) {
         if (this == o) { return true; }
         if (o == null || getClass() != o.getClass()) { return false; }
         final MetricName that = (MetricName) o;
-        return mBeanName.equals(that.mBeanName);
+        return Objects.equals(group, that.group) &&
+            Objects.equals(type, that.type) &&
+            Objects.equals(name, that.name) &&
+            Objects.equals(scope, that.scope) &&
+            Objects.equals(mBeanName, that.mBeanName) &&
+            Objects.equals(tags, that.tags);
     }
 
     @Override
     public int hashCode() {
-        return mBeanName.hashCode();
+        return Objects.hash(group, type, name, scope, mBeanName, tags);
     }
 
     @Override
     public String toString() {
-        return mBeanName;
+        return mBeanName.orElseGet(this::createMBeanName);
```

Comment:
Given your comments, it's probably not worth caching this, right?

Reply:
It should be re-generated on each call rather than cached right?

Reply:
I might have misread your question initially. The current behavior is to re-generate it on each `toString()` call, which I think is probably the desired behavior

Reply:
👍🏻